LayerNorm Support in autodiff:

jjsjann123 opened the pull request:
1. extend autodiff by adding entry for layer_norm in symbolic script, we now use native_layer_norm_backward
2. added backward function `layernorm_double_backward` for `native_layer_norm_backward`, preserves double backward support for LayerNorm in autodiff/ScriptModule
3. added python test to verify autodiff on layer_norm with various configuration of optional tensors; (verify the fix in #49430)

Co-authored-by: Ryan Spring <rspring@nvidia.com>
Co-authored-by: Jie <jiej@nvidia.com>

rdspring1 committed:
LayerNorm Support in autodiff:

1. extend autodiff by adding entry for layer_norm in symbolic script;
2. added python test to verify autodiff on layer_norm with various configuration
   of optional tensors;

Co-authored-by: Ryan Spring <rspring@nvidia.com>
Co-authored-by: Jie <jiej@nvidia.com>

jjsjann123 commented:
This PR also tests fix from #49430. cc'ing @rdspring1

mrshenli reviewed (commented):
Test failures look relevant

jjsjann123 commented:
I saw those failures. I'm pretty occupied at this moment, I'll take a look early next week.

jjsjann123 commented:
hmmm, failures here looks real but I don't have a repro on my machine: https://app.circleci.com/pipelines/github/pytorch/pytorch/259756/workflows/6fcd27bc-6964-422e-8435-ba830bf73354/jobs/10128594

I'm on the same cuda version.... :confused:
I tried to rerun CI with ssh to debug on the same setup, but it doesn't seem to have the right container on the machine...

jjsjann123 commented:
Worked out the CI machine and now I have a repro. debugging now....

jjsjann123 commented:
Phew, the layer_norm kernel issue is sneaky.... I spent sooooo much time looking at autodiff/autograd. :cry: 

@mrshenli I did disable the grad grad test on jitted layer_norm (since we are now using native_layer_norm_backward there, which doesn't have a grad function).

jjsjann123 committed:
disabling grad grad on layer_norm jit test; fixing layer_norm cpu/cuda kernel on overlapping tensor

jjsjann123 commented:
Oops, CI is still red, taking another look.

jjsjann123 committed:
disabling gradgrad test for Transformer in JIT

jjsjann123 commented:
Disabled gradgrad for jit test with layer_norm. Good for review again.

eellison commented:
@jsjjann, I think you're running into https://github.com/pytorch/pytorch/issues/20734 with the grad grad failures. I'm going to try to take a look at that next week, but I'm not sure we can merge this PR before that issue is fixed, because it will lead to errors (as you're running into with the gradgrad tests)

jjsjann123 commented:
I'm not using `view`, but rather composite layer norm kernel instead. So I don't think it's the same issue.

> it will lead to errors (as you're running into with the gradgrad tests)

gradgrad failure I have is pytorch complaining about not getting a backward function for `native_layer_norm_backward`.
Are we suppose to keep  gradgrad support for layer norm in jit as well?

eellison commented:
I'll take a look at this soon, sorry for the delay

ngimel commented on aten/src/ATen/native/cpu/layer_norm_kernel.cpp:
```diff
@@ -280,7 +280,7 @@ void LayerNormBackwardKernelImpl(
       X.scalar_type(), "LayerNormBackwardKernelImpl", [&]() {
         LayerNormBackwardKernelImplInternal<scalar_t>(
-            dY, X, mean, rstd, gamma, M, N, dX, dgamma, dbeta);
+            dY.is_contiguous() ? dY : dY.contiguous(), X, mean, rstd, gamma, M, N, dX, dgamma, dbeta);
```
If you are doing this check here now, can you please remove it from derivatives.yaml to avoid duplication?

ngimel commented on torch/csrc/jit/runtime/symbolic_script.cpp:
```diff
@@ -1014,64 +1014,32 @@ const std::vector<std::string> functions = {
+                # TODO: output_mask should be constructed by requires_grad from input tensor
+                # autodiff doesn't support that yet
+                output_mask = [True, False, False]
+                grad_input, jit_grad_weight, jit_grad_bias = torch.native_layer_norm_backward(grad_output, input, normalized_shape, mean, rstd, weight, bias, output_mask)
```
you never use `jit_grad_weight` and `jit_grad_bias`, so `grad_input, _, _`?

ngimel reviewed (commented):
1) Is this a pessimization for a scripted model unless your fuser is turned on? Looks like it, backward for layer_norm looks slower than it is with autograd in unscripted model. 
2) Scripted models with layer_norm that do double backward now error out? This also looks unfortunate.

jjsjann123 commented:
> 1. Is this a pessimization for a scripted model unless your fuser is turned on? Looks like it, backward for layer_norm looks slower than it is with autograd in unscripted model.

This PR shouldn't change behavior for unscripted models since it updated the behavior in symbolic_script.cpp, which is only used by JIT.

> 2. Scripted models with layer_norm that do double backward now error out? This also looks unfortunate.

Would like to hear more on this one. what's missing now is an entry/implementation for layer_norm_backward in derivatives.yaml. I'm a little nervous that someone might be using double backward with a scripted layernorm and could get very mad when we break it.

ngimel commented:
> This PR shouldn't change behavior for unscripted models since it updated the behavior in symbolic_script.cpp, which is only used by JIT.

I understand that this PR does not change behavior of unscripted models, my question was about scripted models, and it looks like unless nvfuser is turned on, perf of scripted models with layernorm will regress.

ngimel commented:
> Would like to hear more on this one. what's missing now is an entry/implementation for layer_norm_backward in derivatives.yaml. I'm a little nervous that someone might be using double backward with a scripted layernorm and could get very mad when we break it.

Scripting team can comment on what's the contract here, but I'd think that if model can be scripted, whatever you can do with unscripted model you should be able to do with the scripted one, and regressing this is not ideal. cc @suo, @eellison

jjsjann123 commented:
> >This PR shouldn't change behavior for unscripted models since it updated the behavior in symbolic_script.cpp, which is only used by JIT

> I understand that this PR does not change behavior of unscripted models, my question was about scripted models, and it looks like unless nvfuser is turned on, perf of scripted models with layernorm will regress.

I haven't measured bw perf without nvfuser. We are using primitive ops for `grad_bias` & `grad_weight`, which could be tricky when fuser can't handle reduction `_grad_sum_to_size`. This is not optimal but is unfortunately needed for what nvfuser can do at this moment.

For existing behavior, IIUC BW of layer_norm goes through primitive ops in eager as well? We can benchmark this but I'll be surprised if it actually performs well in autograd (since autodiff can't handle layer_norm prior to the PR).
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/layer_norm.cpp#L101
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml#L2092-L2100
https://github.com/pytorch/pytorch/blob/master/tools/autograd/derivatives.yaml#L808-L809
https://github.com/pytorch/pytorch/blob/ddf26816d3ca54ce7f3513f618fac93ce67d06e9/torch/csrc/autograd/FunctionsManual.cpp#L2689-L2791

ngimel commented:
Backward of layer norm, when higher order derivatives are not needed, has a fused kernel to compute parameter gradients https://github.com/pytorch/pytorch/blob/65876d3f5134bba3c6ab5dccd1860ce1b9e43b6d/aten/src/ATen/native/cuda/layer_norm_kernel.cu#L367

jjsjann123 commented:
Ah, I was reading this the wrong way `GradMode::is_enabled() || grads[1].defined() || grads[2].defined()`.

You are totally right about the regression. If we are concerned with that, we can stick to the same fused op for now and do a pass to break layer_norm_backward into two parts so fuser can handle it.

jjsjann123 commented:
FYI, we are also actively working on supporting the composite `layer_norm_backward` in nvfuser. So in the long run it is preferred to have it that way.

ngimel commented:
Yes, we are concerned about scripted models regressions, so let's do the decomposition pass for backward. I'm still not sure about higher order derivatives, I'd also prefer not to break them.

jjsjann123 commented:
Quick note that @rdspring1 is working on adding a autograd support for native_layer_norm_backward.

rdspring1 committed:
Create LayerNorm Double Backward Function

jjsjann123 committed:
Merge remote-tracking branch 'origin/master' into HEAD

jjsjann123 committed:
fixing cherry-picked commit for gradgrad

jjsjann123 committed:
address review comments

jjsjann123 commented:
Cherry-picked @rdspring1 's commit to support double backward. :crossed_fingers: Let's see how CI goes

jjsjann123 commented:
rocm failure seems to be unrelated?

jjsjann123 committed:
Merge remote-tracking branch 'upstream/master' into HEAD

albanD commented on aten/src/ATen/native/cuda/layer_norm_kernel.cu:
```diff
@@ -416,7 +416,7 @@ void LayerNormBackwardKernelImpl(
       X.scalar_type(), "LayerNormBackwardKernelImpl", [&]() {
         LayerNormBackwardKernelImplInternal<scalar_t>(
-            dY, X, mean, rstd, gamma, M, N, dX, dgamma, dbeta);
+            dY.is_contiguous() ? dY : dY.contiguous(), X, mean, rstd, gamma, M, N, dX, dgamma, dbeta);
```
That check is already done in the `contiguous()` call, why do you need to redo-it here?

ngimel replied:
It used to be faster (`is_contiguous` never went through dispatch). These days `.contiguous()` avoids dispatch too in the common case, so this is no longer needed

jjsjann123 replied:
Will update.

ngimel commented:
@jjsjann123 what's the status of this PR? If it's reviewable, it should have better description of how it handles double backward and whether there are expected perf differences, if it's not reviewable, can you put it in "Draft" status?

jjsjann123 commented:
> @jjsjann123 what's the status of this PR? If it's reviewable, it should have better description of how it handles double backward and whether there are expected perf differences, if it's not reviewable, can you put it in "Draft" status?

Performance should be gone now since we wgrad/bgrad are computed inside the kernel.
double backward is not yet supported. I'll look into that. (same story for BN in the other PR I just started: #57321).
I'll mark both PR with Draft.

jjsjann123 committed:
review comments on is_contiguous(); fixing failing CI/merge conflicts

jjsjann123 commented:
oops, my local python version needs to be updated to use py::kw_only.... I'll check for regression tomorrow with a fresh build.

jjsjann123 commented:
A very rough profile: https://github.com/jjsjann123/burn_before_reading/blob/main/layer_norm/ln_profile.py

end-2-end runtime with L1 grad is about the same with differentiable graph slightly slower. (0.008846903219819069 vs 
0.008593639358878136) I assume those are overheads for autodiff. (I also checked the profile print, both execute the same kernel except now everything goes through autodiff instead of autograd.

If you look at the API profile when gradgrad is required:
1. Before the changes, L1 grad goes through indefinitely differentiable layernorm, which consists of primitive ops, which runs a lot slower. https://github.com/jjsjann123/burn_before_reading/blob/5f647f000d2a8e9641cd3a8b7a1d7f3bf86ec20f/layer_norm/ln_profile_log_before#L2-L65
2. After the changes, L1 goes through native_layer_norm_backward, and L2 goes through Ryan's gradgrad function
https://github.com/jjsjann123/burn_before_reading/blob/5f647f000d2a8e9641cd3a8b7a1d7f3bf86ec20f/layer_norm/ln_profile_log_after#L8
https://github.com/jjsjann123/burn_before_reading/blob/5f647f000d2a8e9641cd3a8b7a1d7f3bf86ec20f/layer_norm/ln_profile_log_after#L18

jjsjann123 commented:
I think this one is good for review now

jjsjann123 commented:
Any update?

jjsjann123 committed:
Merge remote-tracking branch 'upstream/master' into HEAD

Krovatkin commented on test/test_jit.py:
```diff
@@ -10565,6 +10565,89 @@ def addmm_grad_test(b, x, w):
         self.assertEqual(w.grad, w_ref.grad)
         self.assertEqual(b.grad, b_ref.grad)
 
+    def test_layer_norm_grad(self):
```
1. Should we confirm that layer_norm was indeed auto differentiated?
Also
this test looks very similar to the tests `jit_metaprogramming_utils.py` should be doing:
* fwd result check against ref
* grad check against grad_ref
* grad grad check against grad grad ref

Is there a reason for a separate test?

jjsjann123 replied:
IIRC, the existing test in `jit_metaprogramming_utils.py` doesn't check for all the combination of `weight` / `bias` / `training`. We have this test for better coverage.

But you are absolutely right that there's no good reason to keep them. I can try to come up with a similar test using the meta test infrastructure.

jjsjann123 replied:
I double checked the `jit_metaprogramming_utils.py`, layer_norm does seem to get a good combination of weight/bias there. The problem I mentioned earlier only applies to batch_norm. I'll remove the test here.

Krovatkin reviewed (commented):
a few questions on the jit part

albanD commented on aten/src/ATen/native/cpu/layer_norm_kernel.cpp:
```diff
@@ -282,7 +282,7 @@ void LayerNormBackwardKernelImpl(
       X.scalar_type(), "LayerNormBackwardKernelImpl", [&]() {
         LayerNormBackwardKernelImplInternal<scalar_t>(
-            dY, X, mean, rstd, gamma, M, N, dX, dgamma, dbeta);
+            dY.contiguous(), X, mean, rstd, gamma, M, N, dX, dgamma, dbeta);
```
Why is this needed now?

jjsjann123 replied:
This is because of autodiff from JIT.

Back then we have autograd calling contiguous on grads[0] when it is not contiguous. But that's not visible to autodiff and causes breakage. We want to keep things consistent, so we are now calling contiguous here and removed the corresponding part in autodiff (https://github.com/pytorch/pytorch/pull/50467/files#diff-e4c2f99a2404e98c3586e07425da73008f36b1bada790648a7297af141d37f8cR975)

albanD replied:
Ho ok, It was hard to see what was removed from that formula :D

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -972,7 +972,12 @@
-  input, weight, bias: "GradMode::is_enabled() || grads[1].defined() || grads[2].defined() ? infinitely_differentiable_native_layer_norm_backward(grads[0], grads[1], grads[2], input, result1, result2, weight, normalized_shape, eps, grad_input_mask) : (grads[0].defined() ? native_layer_norm_backward(grads[0].is_contiguous() ? grads[0] : grads[0].contiguous(), input, normalized_shape, result1, resu ...
+  input, weight, bias: "GradMode::is_enabled() || grads[1].defined() || grads[2].defined() ? infinitely_differentiable_native_layer_norm_backward(grads[0], grads[1], grads[2], input, result1, result2, weight, normalized_shape, eps, grad_input_mask) : (grads[0].defined() ? native_layer_norm_backward(grads[0], input, normalized_shape, result1, result2, weight, bias, grad_input_mask) : std::tuple<Te ...
+
+- name: native_layer_norm_backward(Tensor grad_out, Tensor input, int[] normalized_shape, Tensor mean, Tensor rstd, Tensor? weight, Tensor? bias, bool[3] output_mask) -> (Tensor, Tensor, Tensor)
```
I'm not sure how this can ever be called as the code above is explicitly using `infinitely_differentiable_native_layer_norm_backward` if GradMode is enabled (meaning that you're preparing for double backward).

jjsjann123 replied:
Same as the comment above. In jit, autodiff doesn't use autograd logic. So in the case of double backward, first backward call would use `native_layer_norm_backward` as the backward function defined in `symbolic_script.cpp` (https://github.com/pytorch/pytorch/pull/50467/files#diff-e690d44b82baeb013fa244ea3ceb4b2b9bf7a9b7b525723818652186e6af2468R1156)

The second backward function kicks in and it will be looking for a backward function for `native_layer_norm_backward`

albanD replied:
That sounds bad no?
If we have an efficient implementation of the backward that is infinitely differentiable, we should use that for both autograd.

albanD replied:
Note that using both of them will also allow the existing gradcheck for the vanilla autograd to also check this implementation to make sure that it is correct.

jjsjann123 replied:
The meta test case also does grad & gradgrad check, which in the jitted version covers the added code path.

> That sounds bad no?
It does sounds bad. The alternative here is to remove `infinitely_differentiable_native_layer_norm_backward` and just use `native_layer_norm_backward`.

albanD replied:
> The alternative here is to remove infinitely_differentiable_native_layer_norm_backward and just use native_layer_norm_backward.

Yes! Code removal 🎉

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -3043,6 +3043,146 @@ std::tuple<Tensor, Tensor, Tensor> batchnorm_double_backward(
+  return src_expanded;
+}
+
+std::tuple<Tensor, Tensor, Tensor, Tensor> layernorm_double_backward(
```
nit: layer_norm to be consistent with other functions

jjsjann123 replied:
Will do

jjsjann123 committed:
Merge remote-tracking branch 'upstream/master' into HEAD

jjsjann123 committed:
1. removing infinitely_differentiable_native_layer_norm_backward
2. enabling native_layer_norm_backward for gradgrad in eager
3. renaming layernorm_double_backward to layer_norm_double_backward

jjsjann123 committed:
python test removed

jjsjann123 commented:
Oops, looks like there's wrong result now.... need to double check the kernel implementation. :stuck_out_tongue:

albanD commented:
@jjsjann123 by the way, that means that the manual test that was added for the jit case was not covering all the cases!

jjsjann123 commented:
Just realized that eager infinitely differentiable case also handles `dmean` & `drstd`, which is not handled by `native_layer_norm_backward`, nor the added `native_layer_norm_double_backward`.

I think that's causing the wrong result that we are seeing in `gradgradcheck`. But it seems quite mysterious to me why that's not causing error for gradgradcheck for `native_batch_norm`, which follows about the same math (except the broadcasting semantics for affine weight) and is implemented with similar code.

It's also quite strange that jit meta test `python test_jit.py TestJitGeneratedFunctional.test_nn_layer_norm` runs fine when we keep `infinitely_differentiable_layer_norm_backward` for eager. I thought we have warm up runs in the test, so it should hit the double backward kernel here.

@rdspring1 is working on getting `dmean` & `drstd` into the backward function, we'll see how that goes then.

jjsjann123 committed:
WIP

jjsjann123 committed:
cherry-picking Ryan's fix in https://github.com/jjsjann123/pytorch/pull/4

jjsjann123 committed:
build fix

jjsjann123 committed:
Merge remote-tracking branch 'upstream/master' into HEAD

jjsjann123 committed:
cherry-picking Ryan's fix on gI for double layernorm

jjsjann123 committed:
code cleaning

jjsjann123 commented:
Thanks to @rdspring1 's fix on double backward, CI was all green. The PR is good for another review now @albanD @Krovatkin

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -1010,7 +1010,13 @@
+
+- name: native_layer_norm_backward(Tensor grad_out, Tensor input, int[] normalized_shape, Tensor mean, Tensor rstd, Tensor? weight, Tensor? bias, bool[3] output_mask) -> (Tensor, Tensor, Tensor)
+  input, weight, grad_out: layer_norm_double_backward(input, weight, grads[0], grads[1], grads[2], grad_out, mean, rstd, normalized_shape, grad_input_mask)
+  bias: Tensor()
```
The bias value is not used when evaluating native_layer_norm_backward ? We can remove it then to save memory

jjsjann123 replied:
This one is a little messy.
bias is only used to get dbias options dtype: https://github.com/pytorch/pytorch/blob/289bcac618448b04bd68a2acd577c0261e368da6/aten/src/ATen/native/cuda/layer_norm_kernel.cu#L555

Technically weight & bias should have the same dtype. But weight/bias are independent and weight could be undefined.
I can remove bias and replace it with an optional dtype.

albanD replied:
Yes replacing it with dtype is better as the codegen will stop saving the full bias Tensor for no reason.

jjsjann123 replied:
Do we want to do that in this PR as well?
This one has been dragging for a while and I would really like to have the two autodiff PRs merged sooner than later.

I can start another PR for the API change at the same time.

jjsjann123 replied:
Turns out it's quite sticky to update that in derivatives.yaml: https://github.com/pytorch/pytorch/issues/62934

albanD reviewed (commented):
Looks mostly ok.
Did you run some benchmark to see how the new implementation compares to the previous one. Both when doing a single backward and when preparing for double backward?

jjsjann123 commented:
> Looks mostly ok.
> Did you run some benchmark to see how the new implementation compares to the previous one. Both when doing a single backward and when preparing for double backward?

Will do.

jjsjann123 commented:
![speedup_ln_backward](https://user-images.githubusercontent.com/3709243/128099035-a39df738-401c-499d-ae0b-c1b0d20f8fc5.png)

I'm only posting the speed up comparing the old and new implementation.
1. single backward time is about the same, this is as expected since we didn't change a single thing there.
with double backward:
2.1 we see major speedup for first backward time (2.6x - 5x), since we are now using the fused version instead of primitive ops.
2.2. second backward function shows slight speedup and less than 3% regression on worst cases.

jjsjann123 commented:
I'm using this simple script:
```
import torch
import time

device = "cpu"
ITER = 500

m_size = (16, 32, 64, 128, 256, 512)
n_size = (16, 32, 64, 128, 256, 512)

for m in m_size:
  for n in n_size:
    x = torch.randn(m, n, requires_grad=True, device=device)
    x.retain_grad()
    w = torch.ones(n, requires_grad=True, device=device).to(x.dtype)
    b = torch.zeros(n, requires_grad=True, device=device).to(x.dtype)
    
    o = torch.native_layer_norm(x, (n,), w, b, 1e-5)
    grad = torch.rand_like(x).requires_grad_()

    backward_time = 0
    for i in range(ITER):
      o = torch.native_layer_norm(x, (n,), w, b, 1e-5)
      x.retain_grad()
      x.grad = None
      w.grad = None
      b.grad = None
      grad.grad = None

      torch.cuda.synchronize()
      start = time.time()
      grad_inputs = torch.autograd.grad(o[0], (x, w, b), grad)
      torch.cuda.synchronize()
      end = time.time()
      backward_time += end - start

    print("input size: {}x{}, iter times: {}, single backward time per iteration: {} ms".format(m, n, ITER, backward_time / ITER * 1000)) 

for m in m_size:
  for n in n_size:
    x = torch.randn(m, n, requires_grad=True, device=device)
    x.retain_grad()
    w = torch.ones(n, requires_grad=True, device=device).to(x.dtype)
    b = torch.zeros(n, requires_grad=True, device=device).to(x.dtype)
    
    o = torch.native_layer_norm(x, (n,), w, b, 1e-5)
    grad = torch.rand_like(x).requires_grad_()

    backward_time = 0
    double_backward_time = 0
    for i in range(ITER):
      o = torch.native_layer_norm(x, (n,), w, b, 1e-5)
      x.retain_grad()
      x.grad = None
      w.grad = None
      b.grad = None
      grad.grad = None

      torch.cuda.synchronize()
      start = time.time()
      grad_inputs = torch.autograd.grad(o[0], (x, w, b), grad, create_graph=True, allow_unused=True)
      torch.cuda.synchronize()
      end = time.time()
      backward_time += end - start
      gg = None
      for grad_i in grad_inputs:
        g = grad_i.sum()
        if gg is None:
          gg = g
        else:
          gg = gg + g
      torch.cuda.synchronize()
      start = time.time()
      grad_inputs2 = torch.autograd.grad(gg, (x, w, b, grad), allow_unused=True)
      torch.cuda.synchronize()
      end = time.time()
      double_backward_time += end - start

    print("input size: {}x{}, iter times: {}, backward time per iteration: {} ms".format(m, n, ITER, backward_time / ITER * 1000)) 
    print("input size: {}x{}, iter times: {}, double backward time per iteration: {} ms".format(m, n, ITER, double_backward_time / ITER * 1000)) 
```

The benchmark comparison of old implementation was using this commit: https://github.com/jjsjann123/pytorch/tree/autodiff_layer_norm_benchmark

jjsjann123 committed:
code cleaning

jjsjann123 committed:
Merge remote-tracking branch 'upstream/master' into HEAD

albanD reviewed (approved):
LGTM

Note that I'm by no mean a jit symbolic diff expert so I might have missed things there.

Krovatkin reviewed (approved):
:shipit:

eellison commented:
@albanD @jjsjann123 is there anything that we think could change numerics here ?

jjsjann123 commented:
One thing that we changed is dropping `dmean` and `drstd`.... :face_with_head_bandage:
https://github.com/jjsjann123/pytorch/blob/dd071116629ac1b6bedea1fd5800d74ae660a09f/tools/autograd/derivatives.yaml#L1012-L1013
But batch_norm does the same thing...
